src/graph.jl

```diff
@@ -64,42 +66,119 @@ function random_walk(g::WeightedAdjacencyMatrix{T}, x::Vector{T}, steps::Int64)
     return x
 end
 
-mutable struct AdjacencyList <: AbstractGraph
+mutable struct DiAdjacencyList <: AbstractGraph
     relation::Vector{Vector{Int64}}
     n_vertices::Int64
 
-    AdjacencyList(n::Int64) = new([Int[] for i=1:n], n)
+    DiAdjacencyList(n::Int64) = new([Int[] for i=1:n], n)
 end
 
-# TODO: WeightedAdjacencyList
+mutable struct WeightedAdjacencyList{T} <: AbstractGraph
+    relation::Vector{Vector{Int64}}
+    weight::Vector{Vector{T}}
+    n_vertices::Int64
 
-Base.getindex(g::AdjacencyList, i::Int64) = 0 < i < g.n_vertices ? i : -1
+    function WeightedAdjacencyList{T}(n::Int64) where T
+        return new(
+            [Int[] for i=1:n],
+            [T[] for i=1:n],
+            n
+        )
+    end
+end
 
-nv(g::AdjacencyList) = g.n_vertices
+const AdjacencyListLike = Union{DiAdjacencyList, WeightedAdjacencyList}
 
-ne(g::AdjacencyList) = sum([length(g.relation[v]) for v in 1:(g.n_vertices)])
+function Base.getindex(
+    g::AdjacencyListLike,
+    i::Int64
+)
+    return 0 < i < g.n_vertices ? i : -1
+end
 
-neighbor(g::AdjacencyList, v::Int64) = g.relation[v]
+nv(g::AdjacencyListLike) = g.n_vertices
 
-degree(g::AdjacencyList,  v::Int64) = length(g.relation[v])
+function ne(g::AdjacencyListLike)
+    return sum([length(g.relation[v]) for v in 1:(g.n_vertices)])
+end
 
-function Base.show(io::IO, g::AdjacencyList)
-    print(io, "AdjacencyList(")
+function neighbor(
+    g::AdjacencyListLike,
+    v::Int64
+)
+    return g.relation[v]
+end
+
+function degree(
```

Comment:
A definition of degree, mathematically, is the number of edge incident to a vertex. However, a general definition of degree in weighted graph is the sum of edge weight, which is incident to a vertex.